### PR TITLE
Update Gulp to 4.0.0, add package-lock.json

### PR DESCRIPTION
**Pull request by Aaron1011**

### CODE REVIEW

**Comment by phit** [comment]
I was curious if this is recommended and indeed it is https://stackoverflow.com/questions/44206782/do-i-commit-the-package-lock-json-file-created-by-npm-5/44210813#44210813

**Comment by Spongy** [comment]
A preview for this pull request is available at https://cdn.rawgit.com/Spongy/SpongeDocs-PRs/0e838f8/.

Here are some links to the pages that were modified:



_Since the preview frequently changes, please link to [this comment](https://github.com/SpongePowered/SpongeDocs/pull/715#issuecomment-402541356), not to the direct url to the preview._

File: package-lock.json [hunk 1]
```diff
@@ -0,0 +1,4544 @@
+{
+  "name": "SpongeDocs",
+  "version": "0.0.1",
```
Comment (ST-DDT) [inline package-lock.json:3]: Can we increment this version to something more meaningful? Such as the API version.

Comment (phit) [reply package-lock.json:3]: this file is auto generated and is not modified manually, it basically makes a snapshot of all your installed packages

and the version here is pulled from package.json https://github.com/SpongePowered/SpongeDocs/blob/stable/package.json#L3

Comment (ST-DDT) [reply package-lock.json:3]: That is what I meant (update version in `package.json`).

**Comment by ST-DDT** [comment]
If I build this PR locally it will rewrite/change the `package-lock.json` which would cause merge conflicts.

**Comment by ST-DDT** [comment]
I just tried it again on a clean install of Win7 x64:

````bash
$ node -v
v10.6.0
$ npm -v
6.1.0
$ gulp -v
CLI version 2.0.1
Local version 4.0.0
$ python --version
Python 3.7.0
````

The first problem that I encountered was that the `npm install -g gulp` installed gulp `3.9.1`.
After deinstalling it and installing gulp-cli globally I could install gulp 4 and make it work on the command line.

Then once again after running npm install:

````bash
$ npm install
npm WARN optional SKIPPING OPTIONAL DEPENDENCY: fsevents@1.2.4 (node_modules\fsevents):
npm WARN notsup SKIPPING OPTIONAL DEPENDENCY: Unsupported platform for fsevents@1.2.4: wanted {"os":"darwin","arch":"any"} (current: {"os":"win32","arch":"x64"})

added 447 packages from 265 contributors and audited 4895 packages in 10.305s
found 2 vulnerabilities (1 low, 1 critical)
  run `npm audit fix` to fix them, or `npm audit` for details
````

Git reported uncommited changes to me:

````bash
$ git status
On branch npm-update
Your branch is up to date with 'origin/npm-update'.

Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git checkout -- <file>..." to discard changes in working directory)

        modified:   package-lock.json

no changes added to commit (use "git add" and/or "git commit -a")
````

There I got a `"optional"=true for every dependency.

````diff
         "number-is-nan": {
           "version": "1.0.1",
           "bundled": true,
-          "dev": true
+          "dev": true,
+          "optional": true
         },
````

I assume that the `optional` stuff is something from node 10. If that is indeed the case we should explititly state that we use node 8 on the Readme. And also need to update the the gulp install command from `npm install -g gulp` to `npm intall -g gulp-cli` (I prefer this) or `npm install -g gulp@4.0.0`.

------------------

Next attempt using node 8:

````bash
$ node -v
v8.11.3
$ npm -v
5.6.0
$ gulp -v
CLI version 2.0.1
Local version 4.0.0
$ python --version
Python 3.7.0
````

I got the following changes in my `package.json` and `package-lock.json` file:

````diff
       "dev": true,
       "requires": {
-        "mime-types": "~2.1.18",
+        "mime-types": "2.1.18",
         "negotiator": "0.6.1"
       }
     },
@@ -20,7 +20,7 @@
       "integrity": "sha512-SFKX67auSNoVR38N3L+nvsPjOE0bybKTYbkf5tRvushrAPQ9V75huw0ZxBkKVeRU9kqH3d6HA4xTckbwZ4ixmA==",
       "dev": true,
       "requires": {
-        "ansi-wrap": "^0.1.0"
+        "ansi-wrap": "0.1.0"
       }
     },
     "ansi-gray": {
@@ -56,8 +56,8 @@
       "integrity": "sha512-5teOsQWABXHHBFP9y3skS5P3d/WfWXpv3FUpy+LorMrNYaT9pI4oLMQX7jzQ2KklNpGpWHzdCXTDT2Y3XGlZBw==",
       "dev": true,
       "requires": {
-        "micromatch": "^3.1.4",
-        "normalize-path": "^2.1.1"
+        "micromatch": "3.1.10",
+        "normalize-path": "2.1.1"
       }
     },
     "append-buffer": {
````

Is this an error on my side? And if yes how do I fix it?

-------------------

We should later check the `2 vulnerabilities` found by npm 6.

**Comment by ST-DDT** [comment]
@Aaron1011 Status?

**Comment by Aaron1011** [comment]
@ST-DDT: I'm not extremely familiar with NPM, so I'm not sure why your local`package-lock.json` would be getting modified.

**Comment by ST-DDT** [comment]
Which of version node and npm  are you using?
Can anybody else confirm this behaviour?

**Comment by Shinigami92** [comment]
I'll try to explain that to you
node has nothing to do with the package-lock.json!
it is purely npm

Since npm 6+, a more specific trigger has been added to the package-lock.json, so now ^ and ~ are included

currently npm 6.1 (alternatively 6.2) and node 8.11.3 (LTS) should be used
therefore install node 8.11.3 and then upgrade from npm 5.6.0 with `npm install npm@6.1.0 -g` (or npm `install npm@latest -g`)

npm 6.1+ has the following two advantages
- [npm audit](https://docs.npmjs.com/cli/audit) to check vulnerabilities
- [npm ci](https://docs.npmjs.com/cli/ci) for continuous integration

you should also add [engines](https://docs.npmjs.com/files/package.json#engines) to your package.json and change the name from "SpongeDocs" to "sponge-docs"
final package.json looks like this:
```json
{
  "name": "sponge-docs",
  "version": "0.0.1",
  "private": true,
  "devDependencies": {
    "del": "^3.0.0",
    "gulp": "^4.0.0",
    "gulp-webserver": "^0.9.1"
  },
  "engines": {
    "npm": ">=6.1.0",
    "node": ">=8.11.3"
  }
}
```

Further tips:

1. for hobby projects you can use ^
   you should rather use ~ for business projects
2. do a clean install if you update a dependency manually
   `"scripts": { "clean:install": "rm -Rf node_modules package-lock.json && npm install" }`

**Comment by ST-DDT** [comment]
Well it works on one on my devices as well when following the above suggestions so I guess its just something in the config of the other system.
As for the other suggestions we might as well move them to a separate PR for a cleaner history unless anybody wants to address them directly.

**Review by ST-DDT — APPROVED** [review]

**Pull request merged**